### DISCUSSION TITLE
Rendering Objects while Training a Model

### DISCUSSION THREAD

**Discussion by royvelich** [pygfx/pygfx#572] [Q&A]
Hi,
I would like to use your package as a render engine for 3D models that my Graph Neural Network produces.
Therefore, I would like to run the rendering window in parallel with my training loop, so I can update my rendered objects every few training batches, and see how the outputs of my network progress over time.
Where can I find a minimal example that shows how to run the rendering window in parallel to the main process, which also allows to update its objects dynamically?

Thanks,
Roy

**Comment by royvelich** [comment]
Anything?

**Reply by Korijn** [reply]
It's quite a tall order actually, we need some time to think about it. You're asking for a pretty complicated thing, and it takes some brain power to come up with a simple, minimal example. Give us some time.

**Reply by royvelich** [reply]
I understand. I really appreciate your time and efforts. I didn't realize that it might be that complicated.

**Comment by kushalkolar** [comment]
I'm not sure on the details of what you're trying to do, but for example if you're using pytorch to train your models in one process, you can have another process that runs `pygfx` and within an animation function you can poll a multiprocessing `Queue` to receive new data from the pytorch process and update the scene. If things get very complex and inter-process communication isn't a major bottleneck you can use `zmq` instead of Python multiprocessing queues. 

We don't exactly do this, but I do know someone who's using a library built on top of `pygfx` to plot live sensor data using another process.

**Reply by royvelich** [reply]
That sounds like a neat solution for my current requirements. Will try to implement it. Also, I wonder, you wrote "We don't exactly do this", what did you mean by that? what is the alternative way to do it?

Thanks for your time and efforts.

**Reply by kushalkolar** [reply]
I know someone who's using shared memory instead of multiprocessing Queues. But we've found that if you don't need to send more than 2.5GB/s then it's not worth the extra work of managing shaded memory. Multiprocessing Queue should be sufficient for most tasks.

**Comment by almarklein** [comment]
> Where can I find a minimal example that shows how to run the rendering window in parallel to the main process

We haven't done anything like that yet, so there's no example that I know of. 

In principal it should be possible to do what you want. However, pygfx is (most probably) not thread-safe, so you'll want to make sure that only one thread is talking to pygfx. One other possible issue is that I'm not sure how things will work out with the GUI window and events, which you want to handle in that same thread. If you do the viz in the main thread and the training in a separate thread, I presume that part at least will be easier.

So it looks like you're a pioneer in this regard 😉 We'd be very happy to accept a PR with an minimal example that does what you need, to help other users that want to do something similar.

**Reply by royvelich** [reply]
Thanks for the replay. Sure, I will share my PR once it is working.

**Reply by royvelich** [reply]
Ok, I have something working. Can I write the example in the context of training a model with pytorch, or should it be a more generic example that shows how to run and update a rendered scene in a different process?

**Reply by Korijn** [reply]
I guess both would be useful actually! :)

**Reply by royvelich** [reply]
Hi guys, (@Korijn @almarklein @kushalkolar)
I have submitted a pull request with a working example that demonstrates how to train a model and render the results at the same time. Specifically, the example just trains a GNN to predict the Gaussian curvature at each mesh vertex, by overfitting the ground-truth curvature.
